internal/jujuclistore/store.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2025 Canonical.
+
+package jujuclistore
+
+import (
+	"bytes"
+	"context"
+	"html/template"
+	"io"
+	"net/http"
+	"os"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/juju/clock"
+	jujuerrrors "github.com/juju/errors"
+	"github.com/juju/retry"
+	"github.com/juju/zaputil/zapctx"
+	"go.uber.org/zap"
+)
+
+// launchPadURL is the base URL for the Juju binary downloads.
+const launchPadURL = "https://launchpad.net/juju"
+
+// launchPadTemplate is the template for constructing the download URL for Juju binaries.
+const launchPadTemplate = "{{.BaseURL}}/{{.VersionWithMinor}}/{{.VersionWithPatch}}/+download/juju-{{.VersionWithPatch}}-{{.Os}}-{{.Arch}}.tar.xz"
+
+var retryRequestError = jujuerrrors.New("retry request error")
+
+// JujuCLIStoreConfig holds the configuration for the Juju binary fetcher.
+type JujuCLIStoreConfig struct {
+	BaseURL string // Base URL for the Juju binary downloads. Example: "https://launchpad.net/juju"
+	Dir     string // Directory to store the downloaded binaries. Defaults to the system's temp directory.
+
+	MaxEntries int // Maximum number of entries to keep in the directory. Defaults to 2.
+}
+
+// JujuBinarySpec defines the specifications for a Juju binary to be downloaded.
+type JujuBinarySpec struct {
+	VersionWithMinor string // Version with minor version number, e.g., "3.6"
+	VersionWithPatch string // Version with patch version number, e.g., "3.6.2"
+	Os               string // Operating system, e.g., "linux"
+	Arch             string // Architecture, e.g., "amd64"
+}
+
+type jujuCLIStore struct {
+	config   JujuCLIStoreConfig
+	template template.Template
+
+	entries map[string]*Binary // Map to keep track of downloaded entries
+	lock    sync.Mutex         // Protects the entries map
```

Comment:
nit but I've normally seen the lock go above other fields and commented with something like "protects the fields below". But not something that needs to be changed, just sharing.